### ISSUE TITLE
MSCrashesDelegate callback methods are not called on main thread

### ISSUE THREAD

**Issue by hypherion2** [microsoft/appcenter-sdk-apple#1723]
### **Description**

The methods:
- (void)crashes:(MSCrashes *)crashes didSucceedSendingErrorReport:(MSErrorReport *)errorReport

and
 
- (void)crashes:(MSCrashes *)crashes didFailSendingErrorReport:(MSErrorReport *)errorReport withError:(NSError *)error

are called on a different thread.

### **Repro Steps**

Please list the steps used to reproduce your issue.

1. Implement delegates
2. Try to perform UI action from within the delegate methods
3. Crash, not on UI thread.

### **Details**

1. Which SDK version are you using?
    - 2.2.0
2. Which OS version did you experience the issue on?
    - OS X 10.14.5
3. Which Cocoapods version are you using (run `pod --version`)?
    - N/A
4. What device version did you see this error on?  Were you using an emulator or a physical device?
    - Macbook
5. What language are you using?
    - [ x] Objective C
    - [ ] Swift

**Comment by annakocheshkova** [comment]
Hi @hypherion2 indeed, we call these methods from a background. This is expected right now. 
But I see why one would expect them to be called on UI. Let me discuss this with the team.
In any case, we should at least update our docs.

**Comment by AnastasiaKubova** [comment]
The fix has been released in release [2.5.0](https://github.com/microsoft/appcenter-sdk-apple/releases/tag/2.5.0)

**Comment by hypherion2** [comment]
I have tried this fix and I think it introduces another problem, mainly the one described here:
https://www.thecave.com/2015/08/10/dispatch-async-to-main-queue-doesnt-work-with-modal-window-on-mac-os-x/

The _userConfirmationHandler_ is called in a dispatch_sync block. In the handler I display a modal dialog to confirm user wants to send the crash report. After user confirmed, I'm waiting for either _didSucceedSendingErrorReport_ or _didFailSendingErrorReport_ to close the dialog. These never come up because they're called in a dispatch_async block and essentially the app (and SDK thread) is blocked.

At this point, it would be better to revert this fix and just let the app do its own _performSelectorOnMainThread_ or use another approach. Let me know if I should create another issue or this one can be reopened.

**Comment by russelarms** [comment]
@hypherion2 thanks for letting us know, we will discuss this with the team.

**Comment by russelarms** [comment]
Is it possible for you to use code below instead of handling error reports sending inside a modal window?
```
[MSCrashes setUserConfirmationHandler:(^(NSArray<MSErrorReport *> *errorReports) {
               // Use MSAlertViewController to show a dialog to the user where they can choose if they want to provide a crash report.
               NSAlert *alert = [[NSAlert alloc] init];
               [alert setMessageText:@"Sorry about that!"];
               [alert setInformativeText:@"Do you want to send an anonymous crash "
                                         @"report so we can fix the issue?"];
               [alert addButtonWithTitle:@"Always send"];
               [alert addButtonWithTitle:@"Send"];
               [alert addButtonWithTitle:@"Don't send"];
               [alert setAlertStyle:NSWarningAlertStyle];

               switch ([alert runModal]) {
               case NSAlertFirstButtonReturn:
                 [MSCrashes notifyWithUserConfirmation:MSUserConfirmationAlways];
                 break;
               case NSAlertSecondButtonReturn:
                 [MSCrashes notifyWithUserConfirmation:MSUserConfirmationSend];
                 break;
               case NSAlertThirdButtonReturn:
                 [MSCrashes notifyWithUserConfirmation:MSUserConfirmationDontSend];
                 break;
               default:
                 break;
               }

               return YES;
             })];
```

**Comment by hypherion2** [comment]
@russelarms Yes that is possible but in my app I have a more complicated dialog and I do need to wait for confirmation that crash report was sent before closing the dialog and letting the user continue running the app. The point is that with the current implementation that is not possible while with HockeyApp it was.

**Comment by jaelim-ms** [comment]
@hypherion2 Thanks for bring that up. I understand your problem and sounds like there are two options either
1. revert the fix or
2. don't use GCD for those callbacks at least in macOS to support modal window.
I'll discuss this with my team and will get it back to you.

**Comment by jaelim-ms** [comment]
@hypherion2 Do you have any preference for the suggested options? I want to hear your perspective before making a decision.

**Comment by hypherion2** [comment]
It's OK to revert it, can post on main thread from the app level.

**Comment by jaelim-ms** [comment]
Thanks for your feedback. Our callbacks need to be called on main thread, will try to take the second option first. Sounds like opening a modal dialog on macOS and receiving callbacks on main thread is somewhat our SDK should handle it.

**Comment by annakocheshkova** [comment]
Hi there! This issue has been fixed in the [3.0.0 release](https://github.com/microsoft/appcenter-sdk-apple/releases/tag/3.0.0). Please check it and reopen in case there are any issues.
